Add Dyn Provider

jvassev opened the pull request:
Add https://dyn.com/ as provider "dyn".

njuettner commented:
Thanks for the PR :) @jvassev,

we will have a look at your PR, but before we do that could you just fix the unused struct in dyn.go?

-> provider/dyn.go:86:1:warning: versionedClient is unused (deadcode)

njuettner commented:
Also we're looking for maintainers for providers which we don't use, any chances you might want to become of the maintainer of dyn? 

Since you made the first PR it would be nice to have someone having a look if new PR's are coming.

jvassev commented:
Hi Nick,
I fixed the go code and added a basic tutorial.

I am fine with being a Dyn maintainer (though I am not affiliated in any way - just I use that a lot).

Any feed back is welcome. I've made some changes to the root package wrt visibility you may want to have a look at.

Thanks,
Julian

jvassev commented:
Btw how is the release process set up? I need to update the tutorial with an image tag that has the dyn provider in it. I have to take care of this in another commit though.

njuettner commented:
@jvassev thanks for considering to be a maintainer. If you use it a lot even better :). Since we're currently only relying on AWS, we can't test other providers really well if there are changes made. 

After reviewing we'll merge this into a new pre-alpha version.

jvassev commented:
@njuettner np, you did all the heavy lifting making it easy to write new providers :)
I am guessing the new version will be 0.5.0? I will also update the stable/external-dns chart given the new provider and the `--registry`, `--zone-id-filter` flags.

If you cannot review now, don't worry I will be running the code for a few days in may lab and monitor it. I still think I can save some calls (API is rate-limited). I will update you when I am done :)

jvassev commented:
@njuettner I think it's ready for review. I've added a few tests and introduced local caching using the records TTL for the cache entries to reduce number of remote calls.
It's been running for a few days now in my lab.

njuettner commented on pkg/apis/externaldns/types.go:
```diff
@@ -117,7 +121,7 @@ func allLogLevelsAsStrings() []string {
 // ParseFlags adds and parses flags from command line
 func (cfg *Config) ParseFlags(args []string) error {
 	app := kingpin.New("external-dns", "ExternalDNS synchronizes exposed Kubernetes Services and Ingresses with DNS providers.\n\nNote that all flags may be replaced with env vars - `--flag` -> `EXTERNAL_DNS_FLAG=1` or `--flag value` -> `EXTERNAL_DNS_FLAG=value`")
-	app.Version(version)
```
Since it's only used in this package, I see no reason to change it.

jvassev replied:
It is used later when committing zone changes: 
```go
		notes := fmt.Sprintf("Change by external-dns@%s, DynAPI@%s, %s on %s",
			d.AppVersion,
			d.DynVersion,
			time.Now().Format(time.RFC3339),
			h,
		)
```

njuettner commented on provider/dyn.go:
```diff
@@ -0,0 +1,559 @@
+	for _, link := range links {
+
+		// strip the /REST/ prefix as dynect always prepends it
+		link = strings.TrimPrefix(link, "/REST/")
```
/REST/ can also be a global var if it's always the same, what do you think?

jvassev replied:
Done

njuettner commented on provider/dyn.go:
```diff
@@ -0,0 +1,559 @@
+			log.Debugf("Pending changes for zone %s: %+v", zone, errorOrValue(err, &response))
+		}
+
+		h, _ := os.Hostname()
```
check the err

jvassev replied:
Done

njuettner commented:
Apart from that LGTM
